components/support/utils/src/main/java/mozilla/components/support/utils/DownloadUtils.kt

```diff
@@ -40,17 +42,46 @@ object DownloadUtils {
      * This method is largely identical to [android.webkit.URLUtil.guessFileName]
      * which unfortunately does not implement RFC 5987.
      */
-    @JvmStatic
-    fun guessFileName(contentDisposition: String?, url: String?, mimeType: String?): String {
-        val filename = extractFileNameFromUrl(contentDisposition, url)
 
-        // Split filename between base and extension
+    @Suppress("Deprecation")
+    @JvmStatic
+    fun guessFileName(
+        contentDisposition: String?,
+        destinationDirectory: String?,
+        url: String?,
+        mimeType: String?
+    ): String {
+        // Split fileName between base and extension
         // Add an extension if filename does not have one
-        return if (filename.contains('.')) {
-            changeExtension(filename, mimeType)
+        val extractedFileName = extractFileNameFromUrl(contentDisposition, url)
+
+        val fileName = if (extractedFileName.contains('.')) {
+            changeExtension(extractedFileName, mimeType)
         } else {
-            filename + createExtension(mimeType)
+            extractedFileName + createExtension(mimeType)
+        }
+
+        return destinationDirectory?.let {
+            uniqueFileName(Environment.getExternalStoragePublicDirectory(destinationDirectory), fileName)
+        } ?: fileName
+    }
+
+    /**
+     * Checks if the file exists so as not to overwrite one already in the destination directory
+     */
+    fun uniqueFileName(directory: File, fileName: String): String {
```

Comment:
This would behave weirdly for files that don't have an extension if I read this correctly. So, say a file named "test" would result in `test`, `(1).test`, `(2).test` instead of `test`, `test (1)`, `test (2)`, because `substringAfterLast` will return the original value if the delimiter is not found?

Reply:
Since this is a public util method now, we should add some basic tests for it :). Here's an example test for the case I describe above: https://github.com/csadilek/android-components/commit/b1eddbdd5be4c25e841684db6a51a4eea9897fb7

Reply:
Added, thanks for figuring out this test, @csadilek! 😄